R/fair-aaa.R

```diff
@@ -0,0 +1,137 @@
+#' `.post` argument.
+#'
+#' @return A function with one argument, `by`, indicating the data-masked
+#' variable giving the sensitive feature. See the documentation on the
+#' return value of fairness metrics like [demographic_parity()],
+#' [equal_opportunity()], or [equalized_odds()] to learn more about how the
+#' output of this function can be used.
+#'
+#' @examples
+#' data(hpc_cv)
+#'
+#' # `demographic_parity`, among other fairness metrics,
+#' # is generated with `fairness_metric()`:
+#' diff_range <- function(x, ...) {diff(range(x$.estimate))}
+#' demographic_parity_ <-
+#'   fairness_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "demographic_parity",
+#'     .post = diff_range
+#'   )
+#'
+#' m_set <- metric_set(demographic_parity_(Resample))
+#'
+#' m_set(hpc_cv, truth = obs, estimate = pred)
+#'
+#' # the `post` argument can be used to accommodate a wide
+#' # variety of parameterizations. to encode demographic
+#' # parity as a ratio inside of a difference, for example:
+#' ratio_range <- function(x, ...) {
+#'   range <- range(x$.estimate)
+#'   range[1] / range[2]
+#' }
+#'
+#' demographic_parity_ratio <-
+#'   fairness_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "demographic_parity_ratio",
+#'     .post = ratio_range
+#'   )
+#'
+#' @export
+fairness_metric <- function(.fn, .name, .post) {
+  if (rlang::is_missing(.fn) || !inherits_any(.fn, c("metric", "metric_set"))) {
+    rlang::abort("`.fn` must be a metric function or metric set.")
+  }
+  if (rlang::is_missing(.name) || !is_string(.name)) {
+    abort("`.name` must be a string.")
+  }
+  if (rlang::is_missing(.post) || !is_function(.post)) {
+    abort("`.post` must be a function.")
+  }
+
+  function(by) {
+    by_str <- rlang::as_string(rlang::enexpr(by))
+    res <-
+      function(data, ...) {
+        gp_vars <- dplyr::group_vars(data)
+
+        res <- dplyr::group_by(data, {{by}}, .add = TRUE)
+        res <- .fn(res, ...)
+
+        if (length(gp_vars) > 0) {
+          splits <- vctrs::vec_split(res, res[gp_vars])
+          .estimate <- vapply(splits$val, .post, numeric(1), ...)
+        } else {
+          .estimate <- .post(res, ...)
+        }
+
+        if (!rlang::is_bare_numeric(.estimate)) {
+          abort(
+            "`.post` must return a single numeric value.",
+            call = rlang::call2("fairness_metric")
+          )
+        }
+
+        if (length(gp_vars) > 0) {
+          res <- dplyr::group_by(res, !!!dplyr::groups(data), .add = FALSE)
+        }
+
+        res <-
+          dplyr::summarize(
+            res,
+            .metric = .name,
+            !!".by" := by_str,
+            .estimator = .estimator[1],
+            .groups = "drop"
+          )
+
+        res$.estimate <- .estimate
+
+        res
+      }
+    res <- new_class_metric(res, direction = "minimize")
+    attr(res, "by") <- by_str
+    res
+  }
+}
+
+diff_range <- function(x, ...) {
+  diff(range(x$.estimate))
```

Comment:
this feels like a major nitpick, but we can calculate this with `min()` and `max()`

``` r
f1 <- function(x) {
  diff(range(x))
}

f2 <- function(x) {
  max(x) - min(x)
}

x <- rnorm(10)

bench::mark(
  f1(x),
  f2(x)
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 f1(x)        2.67µs   3.03µs   238952.        0B     23.9
#> 2 f2(x)      327.83ns 369.04ns  1965169.        0B    197.
```

Reply:
That's a big ol difference. :) You got it! [a9fc2e0](https://github.com/tidymodels/yardstick/pull/434/commits/a9fc2e08f62fb91e4523a15c15f6a9cf08aae95e)